packages/expect/src/jest-utils.ts

```diff
@@ -452,7 +452,7 @@ export function subsetEquality(object: unknown, subset: unknown, customTesters:
           return undefined
 
         return Object.keys(subset).every((key) => {
-          if (isObjectWithKeys(subset[key])) {
+          if (typeof subset[key] === 'object') {
```

Comment:
I'm not entirely convinced with the fix for `subsetEquality` as well, but the "array" specific infinite recursion is due to `seenReference` is missing array.